proxy/stats_handler.go

```diff
@@ -0,0 +1,243 @@
+package proxy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	jwt "github.com/dgrijalva/jwt-go"
+	"github.com/gorilla/websocket"
+
+	"github.com/rancherio/websocket-proxy/common"
+)
+
+type StatsHandler struct {
+	backend         backendProxy
+	parsedPublicKey interface{}
+}
+
+type statsInfo struct {
+	hostKey     string
+	url         string
+	msgKey      string
+	respChannel <-chan common.Message
+}
+
+func (s *statsInfo) initializeClient(h *StatsHandler) error {
+	if s.hostKey == "" {
+		return fmt.Errorf("hostKey is empty")
+	}
+	msgKey, respChannel, err := h.backend.initializeClient(s.hostKey)
+	if err != nil {
+		return err
+	}
+	s.msgKey = msgKey
+	s.respChannel = respChannel
+	return nil
+}
+
+func (s *statsInfo) closeClient(h *StatsHandler) {
+	h.backend.closeConnection(s.hostKey, s.msgKey)
+}
+
+func (s *statsInfo) connect(h *StatsHandler) error {
+	return h.backend.connect(s.hostKey, s.msgKey, s.url)
+}
+
+func (h *StatsHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	multiHost := false
+
+	if strings.HasSuffix(req.URL.Path, "project") || strings.HasSuffix(req.URL.Path, "project/") || strings.HasSuffix(req.URL.Path, "service") || strings.HasSuffix(req.URL.Path, "service/") {
+		multiHost = true
+	}
+
+	statsInfoStructs, authed := h.auth(req, multiHost)
+	if !authed {
+		http.Error(rw, "Failed authentication", 401)
+		return
+	}
+
+	upgrader := websocket.Upgrader{
+		CheckOrigin: func(r *http.Request) bool { return true },
+	}
+	ws, err := upgrader.Upgrade(rw, req, nil)
+	if err != nil {
+		http.Error(rw, "Failed to upgrade connection.", 500)
+		return
+	}
+
+	var mutex sync.Mutex
+	var countMutex sync.Mutex
+
+	doneCounter := len(statsInfoStructs)
+
+	defer func() {
+		for _, statsInfoStruct := range statsInfoStructs {
+			statsInfoStruct.closeClient(h)
+		}
+		closeConnection(ws)
+	}()
+
+	for _, statsInfoStruct := range statsInfoStructs {
+		err := statsInfoStruct.initializeClient(h)
+		if err != nil {
+			return
+		}
+
+		// Send response messages to client
+		go func(s *statsInfo) {
+			for {
+				message, ok := <-s.respChannel
+				if !ok {
+					return
+				}
+				switch message.Type {
+				case common.Body:
+					mutex.Lock()
+					ws.SetWriteDeadline(time.Now().Add(10 * time.Second))
+					if err := ws.WriteMessage(1, []byte(message.Body)); err != nil {
+						mutex.Unlock()
+						return
```

Comment:
You cannot return here or at line 108. This will leave a channel open with nothing to read it, resulting in a write to that channel blocking forever. They only way to exit this goroutine must be through the `if !ok` block.